compiler/ccgtypes.nim

```diff
@@ -971,19 +982,79 @@ proc isReloadable(m: BModule; prc: PSym): bool =
 proc isNonReloadable(m: BModule; prc: PSym): bool =
   return m.hcrOn and sfNonReloadable in prc.flags
 
+proc parseVFunctionDecl(val: string; name, params, retType: var string; isFnConst, isOverride: var bool) =
+  var afterParams: string
+  if scanf(val, "$*($*)$s$*", name, params, afterParams):
+    isFnConst = afterParams.find("const") > -1
+    isOverride = afterParams.find("override") > -1
+    discard scanf(afterParams, "->$s$* ", retType)
+
+proc parseExistingParams(params:string, parsedParams: var OrderedTable[string, string]) =
+  #instead of messing with all the logic in params we do a pass over them here to "happy" extract them
+  #Ideally genProcParams would be refactored to return a table. So this is a temporary solution.
+  let splitParams = params.multiReplace(("(", ""), (")", "")).split(", ")
+  for p in splitParams:
+    var pair = p.split(" ")
+    if pair.len == 2:
+      parsedParams[pair[0].strip()] = pair[1].strip()
+
+proc genVirtualProcHeader(m: BModule; prc: PSym; result: var Rope; asPtr: bool = false, isFwdDecl : bool = false) =
+  assert sfVirtual in prc.flags
+  # using static is needed for inline procs
+  var check = initIntSet()
+  fillBackendName(m, prc)
+  fillLoc(prc.loc, locProc, prc.ast[namePos], OnUnknown)
+  var typ = prc.typ.n[1].sym.typ
+  var memberOp = "#."
+  if typ.kind == tyPtr:
+    typ = typ[0]
+    memberOp = "#->"
+  let typDesc = getTypeDescWeak(m, typ, check, skParam)
+  var rettype, params: Rope
+  genProcParams(m, prc.typ, rettype, params, check, true, false, true) 
+  var nimParams = {typDesc: "this"}.toOrderedTable #just for ref/consistency. this shouldnt be used (also for obj is (*this))
+  parseExistingParams(params, nimParams)
+  # handle the 2 options for hotcodereloading codegen - function pointer
+  # (instead of forward declaration) or header for function body with "_actual" postfix
+  let asPtrStr = rope(if asPtr: "_PTR" else: "")
+  var name, declParams, userRetTyp: string
+  var isFnConst, isOverride: bool
+  parseVFunctionDecl(prc.loc.r, name, declParams, userRetTyp, isFnConst, isOverride)
+  #first we format the name to get rid of the $
+  template `%`(s: string, args: openArray[string]): string = runtimeFormat(s, args)
+  declParams = declParams.replace("#", "$") % nimParams.pairs.toSeq.mapIt(it[0] & " " & it[1])
+  declParams = declParams.replace("%", "$") % nimParams.keys.toSeq
+  declParams = declParams.replace("^", "$") % nimParams.values.toSeq
```

Comment:
What? Why not process the virtual string pattern like the rest of the compiler does but not shying away from using an index into the string.

Reply:
That wasn't my point. I'm trying to say that a single parsing pass over the pattern that is the argument of `.virtual` and that replaces `%` etc. with the required thing.